domain/secret/service/watcher.go

```diff
@@ -243,170 +150,199 @@ func (w *obsoleteWatcher) init() error {
 
-			// No need to track this one anymore.
-			w.knownSecretURIs.Remove(uri.ID)
+				// No need to track this one anymore.
+				knownSecretURIs.Remove(uri.ID)
+			}
 		}
-	}
-	return currentChanges, nil
-}
-
-func (w *obsoleteWatcher) mergeRevisionChanges(
-	ctx context.Context, currentChanges []string, newChanges []string,
-) ([]string, error) {
-	if len(newChanges) == 0 {
 		return currentChanges, nil
 	}
 
-	// We are receiving all the obsolete revision UUIDs changes in the model, so we need to filter
-	// only the ones that are owned.
-	revisionIDs, err := w.state.GetRevisionIDsForObsolete(
-		ctx, w.appOwners, w.unitOwners, newChanges...,
-	)
-	if err != nil {
-		return currentChanges, errors.Capture(err)
-	}
-
-	return append(currentChanges, revisionIDs...), nil
-}
-
-func (w *obsoleteWatcher) loop() error {
-	defer close(w.out)
-
-	ctx, cancel := w.scopedContext()
-	defer cancel()
-	// To allow the initial event to be sent.
-	out := w.out
-
-	splitEvents := func(events []string) (secretEvents, revisionEvents []string) {
-		if len(events) == 0 {
-			return
+	mergeRevisionChanges := func(currentChanges, newChanges []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		if len(newChanges) == 0 {
+			return currentChanges, nil
 		}
-
-		// The source watcher may emit events from secret_metadata and secret_revision_obsolete tables.
-		// We need to split the events into secret URI strings and revision UUIDs strings.
-		for _, e := range events {
-			if _, err := coresecrets.ParseURI(e); err == nil {
-				secretEvents = append(secretEvents, e)
-				continue
+		// We are receiving all the obsolete revision UUIDs changes in the model, so we need to filter
+		// only the ones that are owned.
+		revisionUUIDs := make([]string, len(newChanges))
+		for i, change := range newChanges {
+			revisionUUIDs[i] = change.Changed()
+		}
+		revisionIDs, err := state.GetRevisionIDsForObsolete(
+			ctx, appOwners, unitOwners, revisionUUIDs...,
+		)
+		if err != nil {
+			return currentChanges, errors.Capture(err)
+		}
+		for _, change := range newChanges {
+			if revisionID, ok := revisionIDs[change.Changed()]; ok {
+				currentChanges = append(currentChanges, newMaskedChangeIDEvent(change, revisionID))
 			}
-			revisionEvents = append(revisionEvents, e)
 		}
-		return
+		return currentChanges, nil
 	}
 
-	var changes []string
-	for {
-		select {
-		case <-w.catacomb.Dying():
-			return w.catacomb.ErrDying()
-		case events, ok := <-w.sourceWatcher.Changes():
-			if !ok {
-				return errors.Errorf("source watcher closed")
-			}
-			secretEvents, revisionEvents := splitEvents(events)
-			var err error
-			changes, err = w.mergeSecretChanges(ctx, changes, secretEvents)
-			if err != nil {
-				return errors.Capture(err)
-			}
-			changes, err = w.mergeRevisionChanges(ctx, changes, revisionEvents)
+	return func(ctx context.Context, changes []changestream.ChangeEvent) (result []changestream.ChangeEvent, err error) {
```

Comment:
Watcher mappers MUST return changes in the order they are given to the mapper. That is to say if the changes received are 1, 2, 3, 4 and the mapper removes event 2 the order returned must be 1, 3, 4.

There is no wiggle room on this. This mapper needs to be re-implemented to take this into account. We will also require a test to assert this behaviour.